tensorflow/lite/kernels/stablehlo_sort.cc

```diff
@@ -0,0 +1,267 @@
+/* Copyright 2024 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <numeric>
+#include <vector>
+
+#include "Eigen/Core"
+#include "tensorflow/lite/builtin_ops.h"
+#include "tensorflow/lite/core/c/builtin_op_data.h"
+#include "tensorflow/lite/core/c/common.h"
+#include "tensorflow/lite/core/subgraph.h"
+#include "tensorflow/lite/kernels/internal/compatibility.h"
+#include "tensorflow/lite/kernels/internal/tensor.h"
+#include "tensorflow/lite/kernels/internal/types.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/kernels/tensor_slice_util.h"
+
+namespace tflite {
+namespace ops {
+namespace builtin {
+namespace stablehlo_sort {
+
+constexpr int kInputTensor = 0;
+constexpr int kOutputTensor = 0;
+
+std::vector<int> FlatToTensorIndex(int flat_index, const int dims,
+                                   const RuntimeShape& shape) {
+  int div = 1;
+  std ::vector<int> index(dims);
+  for (int i = dims - 1; i >= 0; --i) {
+    index[i] = (flat_index / div) % shape.Dims(i);
+    div *= shape.Dims(i);
+  }
+  return index;
+}
+
+template <typename DataType>
+TfLiteStatus EvalImpl(TfLiteContext* context, TfLiteNode* node,
+                      const std::vector<const TfLiteTensor*>& operands,
+                      const int dimension, const bool is_stable,
+                      std::vector<TfLiteTensor*>& results) {
+  const TfLiteStablehloSortParams* data =
+      reinterpret_cast<TfLiteStablehloSortParams*>(node->builtin_data);
+  Subgraph* this_subgraph = reinterpret_cast<Subgraph*>(context->impl_);
+  auto* subgraphs = this_subgraph->GetSubgraphs();
+  Subgraph& comparator_subgraph =
+      *(*subgraphs)[data->comparator_subgraph_index];
+  const int num_inputs = node->inputs->size;
+  const int num_outputs = node->outputs->size;
+
+  const TfLiteTensor* operand = operands[0];
+  const int rank = NumDimensions(operand);
+  const int operand_size = NumElements(operand);
+
+  std::vector<int> operand_index(rank);
+  std::vector<int> output_index(rank);
+  std::vector<DataType> args(operands.size() * 2);
+
+  const int dim_size = GetTensorShape(operand).Dims(dimension);
+  std::vector<int> indices(dim_size);
+  for (int k = 0; k < operand_size; ++k) {
+    auto result_it = FlatToTensorIndex(k, rank, GetTensorShape(operand));
```

Comment:
I think it would probably be more efficient to use [NextIndex](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/types.h#L159-L180) for this.